filecoin-hashers/src/poseidon_types.rs

```diff
@@ -26,55 +30,71 @@ lazy_static! {
+    pub static ref POSEIDON_MD_CONSTANTS_VESTA: PoseidonConstants::<Fq, PoseidonMDArity> =
+        PoseidonConstants::new();
+    pub static ref POSEIDON_CONSTANTS: ShareMap = {
+        let mut tm = ShareMap::custom();
 
-pub trait PoseidonArity: Arity<Fr> + Send + Sync + Clone + Debug {
-    #[allow(non_snake_case)]
-    fn PARAMETERS() -> &'static PoseidonConstants<Fr, Self>;
-}
+        tm.insert::<FieldArity<Fr, U2>>(&*POSEIDON_CONSTANTS_2);
+        tm.insert::<FieldArity<Fr, U4>>(&*POSEIDON_CONSTANTS_4);
+        tm.insert::<FieldArity<Fr, U8>>(&*POSEIDON_CONSTANTS_8);
+        tm.insert::<FieldArity<Fr, U11>>(&*POSEIDON_CONSTANTS_11);
 
-impl PoseidonArity for U0 {
-    fn PARAMETERS() -> &'static PoseidonConstants<Fr, Self> {
-        unreachable!("dummy implementation, do not ever call me")
-    }
-}
+        tm.insert::<FieldArity<Fp, U2>>(&*POSEIDON_CONSTANTS_2_PALLAS);
+        tm.insert::<FieldArity<Fp, U4>>(&*POSEIDON_CONSTANTS_4_PALLAS);
+        tm.insert::<FieldArity<Fp, U8>>(&*POSEIDON_CONSTANTS_8_PALLAS);
+        tm.insert::<FieldArity<Fp, U11>>(&*POSEIDON_CONSTANTS_11_PALLAS);
 
-impl PoseidonArity for U2 {
-    fn PARAMETERS() -> &'static PoseidonConstants<Fr, Self> {
-        &*POSEIDON_CONSTANTS_2
-    }
-}
+        tm.insert::<FieldArity<Fq, U2>>(&*POSEIDON_CONSTANTS_2_VESTA);
+        tm.insert::<FieldArity<Fq, U4>>(&*POSEIDON_CONSTANTS_4_VESTA);
+        tm.insert::<FieldArity<Fq, U8>>(&*POSEIDON_CONSTANTS_8_VESTA);
+        tm.insert::<FieldArity<Fq, U11>>(&*POSEIDON_CONSTANTS_11_VESTA);
 
-impl PoseidonArity for U4 {
-    fn PARAMETERS() -> &'static PoseidonConstants<Fr, Self> {
-        &*POSEIDON_CONSTANTS_4
-    }
+        tm
+    };
 }
 
-impl PoseidonArity for U8 {
-    fn PARAMETERS() -> &'static PoseidonConstants<Fr, Self> {
-        &*POSEIDON_CONSTANTS_8
-    }
-}
+// Used as the key to lookup Poseidon constants for a field `F` and arity `A`.
+pub struct FieldArity<F, A>(PhantomData<(F, A)>)
+where
+    F: PrimeField,
+    A: PoseidonArity<F>;
 
-impl PoseidonArity for U11 {
-    fn PARAMETERS() -> &'static PoseidonConstants<Fr, Self> {
-        &*POSEIDON_CONSTANTS_11
-    }
+impl<F, A> typemap::Key for FieldArity<F, A>
+where
+    F: PrimeField,
+    A: PoseidonArity<F>,
+{
+    type Value = &'static PoseidonConstants<F, A>;
 }
 
-impl PoseidonArity for U16 {
-    fn PARAMETERS() -> &'static PoseidonConstants<Fr, Self> {
-        &*POSEIDON_CONSTANTS_16
-    }
-}
-impl PoseidonArity for U24 {
-    fn PARAMETERS() -> &'static PoseidonConstants<Fr, Self> {
-        &*POSEIDON_CONSTANTS_24
-    }
-}
-impl PoseidonArity for U36 {
-    fn PARAMETERS() -> &'static PoseidonConstants<Fr, Self> {
-        &*POSEIDON_CONSTANTS_36
-    }
-}
+// A marker trait for arities which are in `POSEIDON_CONSTANTS`; we require that 'PoseidonArity<F>`
+// implements `Send + Sync` because those traits are required by `lazy_static`.
+pub trait PoseidonArity<F: PrimeField>: Arity<F> + Send + Sync + Clone + Debug {}
```

Comment:
Note: must make `PoseidonArity` generic over a field `F` because `neptune::Arity` is generic over `F`